content/post/20200426-kubernetes-the-universal-control-plane/index.md

```diff
@@ -0,0 +1,113 @@
+2. **Control Loops**: The core primitive that binds the control and data planes, the loop consists of observing data plane components, determining if stabilizing action needs to be taken, then applying the change back to the data plane. Like the systems they observe, control loops are **continuous**, a feedback mechanism that constantly corrects divergence from the desired state.
+
+<!-- Compare control plane concepts to Kubernetes abstractions; K8s provides abstractions that make controllers simpler to implement; Removing friction means it will be used to alot more; label metadata - undifferentied heavy lifting hard to justify individually -->
+How does Kubernetes apply control plane principles? The Kubernetes API is all about the management of declarative state, how to create/update it and then converging the system to that state. As for control loops, well, it is right there in the [documentation](https://kubernetes.io/docs/concepts/architecture/controller/). Kubernetes innovates on this idea by creating generic machinery to manage all resources as potential control planes. Going back to the idea of first class abstractions, Kubernetes provides components inside Controllers like [Informers and Workqueues](https://engineering.bitnami.com/articles/a-deep-dive-into-kubernetes-controllers.html) that implement the tricky logic for building control loops.
+
+There is an obeservation that removing friction from an action leads to that action being done much more frequently. Kubernetes is headed to that same outcome with control planes. I think we will see control planes used to manage many more kinds of software. Systems are dynamic, and control loops are a powerful way to manage this reality. The bottleneck was always the implementation complexity, and Kubernetes is making moves to make this part tractable through Controllers and Operators paired with the Kubernetes Resource Model.
+
+## Standards Over Implementations
+
+<!-- Standardization = leverage and distribution; Implementing principles of programmable infrastructure and declarative APIs; KRM and controllers are core to the architecture, not married to the implementation; Kubernetes without computation is a schema database; Standardizing the core ideas will allow K8s to evolve the different infra needs across the industry -->
+The common theme threading through this essay is the idea of standardization. The Kubernetes Resource Model leans heavily on enforcing consistency in how resources are structured and tightly constrains the allowed access paths to the data. In return for the limitations on the data model [1] the ecosystem gains leverage, able to use the same primitives across many different abstractions. The primitives also implement the core principles to cloud computing: Resources are layered to create higher level abstractions (like how Deployments abstract Pods), and Controllers allow resources to be declarative by pushing the implementation complexity to the underlying code.
```

Comment:
this comma break sounds awkward: "the ecosystem gains leverage, able to us"

Reply:
Don't use the word essay in the first sentence. It felt a bit jarring like breaking the 4th wall. I'd go with something like "The common theme threading together these decisions is the idea of standardization"